### ISSUE TITLE
Should `happy-1.21.0` be revised to become unbuildable?

### ISSUE THREAD

**Issue by andreasabel** [simonmar/happy#215]
See image in https://github.com/simonmar/happy/issues/214#issue-1051374134:
`happy-1.21.0` is listed on `matrix.hackage` as __FAIL(pkg)__ and I read somewhere that it is deprecated.  Should it be revised to actually be unbuildable (e.g. by setting `base < 0`)?  This would switch it to __OK(no-ip)__.

**Comment by Ericson2314** [comment]
Could we revise it to make bootstrap on by default? If so, that should also make it succeed?

**Comment by simonmichael** [comment]
I'd welcome either of these. The release of happy 1.21 broke my CI, and it wasn't too easy to discover the cause.

**Comment by Ericson2314** [comment]
> Errors
>
> Cannot change the default of a manual flag

No dice

**Comment by int-index** [comment]
The bootstrap does not support all features, so that might have broken some builds as well. What about the `base < 0` trick, would it work?

**Comment by andreasabel** [comment]
> What about the base < 0 trick, would it work?

Yes, this would make the happy-1.21.0 unbuildable, and prune it from the `cabal` constraint solving process.
This was my suggestion.
Seems like changing the flag value would warrant a release of 1.21.0.1 or 1.21.1.

**Comment by int-index** [comment]
> Yes, this would make the happy-1.21.0 unbuildable, and prune it from the cabal constraint solving process.

But it’s marked as deprecated, shouldn’t it already be pruned?

**Comment by andreasabel** [comment]
There are exceptions:
> If all the available versions of a package are non-preferred or deprecated, cabal-install will treat this the same as if none of them are. 


Btw, it does install fine:
```
$ cabal install happy-1.21.0
Resolving dependencies...
Build profile: -w ghc-9.2.1 -O1
In order, the following will be built (use -v for more details):
 - happy-1.21.0 (exe:happy) (requires build)
Configuring executable 'happy' for happy-1.21.0..
Preprocessing executable 'happy' for happy-1.21.0..
Building executable 'happy' for happy-1.21.0..
[ 1 of 23] Compiling AbsSyn      
((SUCCEEDS))
```

**Comment by phadej** [comment]
deprecation is indeed a *preference*, solver can still pick it if it's forced to. `base <0` would be a **hard (unsatisfiable) constraint**.

TL;DR deprecations on Hackage do very little.
IMO not worth using at all
(There are exceptions, but they could be pruned by being able to do a bit more precise revisions).

**Comment by int-index** [comment]
OK, I revised it to specify `base <0`

**Comment by andreasabel** [comment]
See https://hackage.haskell.org/package/happy-1.21.0/revisions/
